scipy/stats/_distribution_infrastructure.py

```diff
@@ -3618,62 +3682,99 @@ def _cdf2_subtraction(self, x, y, **params):
         pmf_x = np.where(x_ == x, self._pmf_dispatch(x_, **params), 0)
         return cdf_ymx + pmf_x
 
+    def _cdf2_subtraction_safe(self, x, y, **params):
+        x_, y_ = np.floor(x), np.floor(y)
+        cdf_ymx = super()._cdf2_subtraction_safe(x_, y_, **params)
+        pmf_x = np.where(x_ == x, self._pmf_dispatch(x_, **params), 0)
+        return cdf_ymx + pmf_x
+
     def _logcdf2_subtraction(self, x, y, **params):
         x_, y_ = np.floor(x), np.floor(y)
         logcdf_ymx = super()._logcdf2_subtraction(x_, y_, **params)
         logpmf_x = np.where(x_ == x, self._logpmf_dispatch(x_, **params), -np.inf)
         return special.logsumexp([logcdf_ymx, logpmf_x], axis=0)
 
+    def _logcdf2_subtraction_safe(self, x, y, **params):
+        x_, y_ = np.floor(x), np.floor(y)
+        logcdf_ymx = super()._logcdf2_subtraction_safe(x_, y_, **params)
+        logpmf_x = np.where(x_ == x, self._logpmf_dispatch(x_, **params), -np.inf)
+        return special.logsumexp([logcdf_ymx, logpmf_x], axis=0)
+
     def _ccdf2_addition(self, x, y, **params):
         a, _ = self._support(**params)
         ccdf_y = self._ccdf_dispatch(y, **params)
         _cdf, args = _kwargs2args(self._cdf_dispatch, kwargs=params)
         cdf_xm1 = _lazywhere(x - 1 >= a, [x - 1] + args, _cdf, fillvalue=0)
-        return ccdf_y + cdf_xm1
+        res = np.asarray(ccdf_y + cdf_xm1)
+        res[np.isnan(x)] = np.nan
+        return res[()]
 
     def _logccdf2_addition(self, x, y, **params):
         a, _ = self._support(**params)
         logccdf_y = self._logccdf_dispatch(y, **params)
         _logcdf, args = _kwargs2args(self._logcdf_dispatch, kwargs=params)
         logcdf_xm1 = _lazywhere(x - 1 >= a, [x - 1] + args, _logcdf, fillvalue=-np.inf)
-        return special.logsumexp([logccdf_y, logcdf_xm1], axis=0)
+        res = np.asarray(special.logsumexp([logccdf_y, logcdf_xm1], axis=0))
+        res[np.isnan(x)] = np.nan
+        return res[()]
 
     def _icdf_inversion(self, x, **params):
-        res = self._solve_bounded(self._cdf_dispatch, x, params=params, xatol=1)
-        res = np.where(res.fl >= 0, np.floor(res.xl), np.floor(res.xr))
+        # For discrete distributions, icdf(p) is defined as the minimum x
+        # such that cdf(x) >= p.
+        # np.nextafter(xatol
+        res = self._solve_bounded(self._cdf_dispatch, x, params=params, xatol=0.9)
+        res = np.where(res.fun >= 0, np.floor(res.x), np.ceil(res.x))
+        # If cdf(res) < x, then the true result must be one greater.
+        res = np.where(self.cdf(res) >= x, res, res + 1.0)
```

Comment:
Are we guaranteed to stay within the support?

Reply:
Maybe I can answer that question: since `res` is an integer, `res + 1.0` can only be greater than `b` (right end of the support) if `res = b`. Therefore, `cdf(res) = 1 >= x` would be satisfied, so this would just return `res`.

That said, we shouldn't call the public `cdf` method here; use `_cdf_dispatch`.

Reply:
I'm trying to avoid *two* `where`s and an extra CDF call. 
Forget about `res.x` for a moment. We know that `xl` and `xr` bound the True `x`. 

If we knew the true `x`, we would always want `ceil(x)`, right?

`x <= xr < x + 1`, so mustn't `ceil(x)` equal either `floor(xr)` or `floor(xr) + 1`?

In that case, can we get rid of the first `where`?

Reply:
We discussed that `atol` really does control the width of `res.xr - res.xl`. Because `res.x` is either `res.xr` or `res.xl`, this controls the x-distance between `res.x` and the true value to `atol`, too. Another possibility was to double the tolerance and manually take `x = (res.xl + res.xr)/2` to use instead of `res.x` in the code as shown. This _could_ reduce the number of iterations required by 1, in which case what you have here might be preferable to the suggestion. However, it's not guaranteed. I'd prefer to either study whether this would reduce the number of iterations in practice or to use https://github.com/mdhaber/scipy/pull/121#discussion_r2059335613.